src/html.js

```diff
@@ -80,14 +80,35 @@ export function HTMLString(str) {
 	this.value = str;
 }
 
-// adapted from TiddlyWiki <http://tiddlywiki.com> and Python 3's `html` module
+const HTML_ENCODE_REPLACE = {
+	"&": "&amp;",
+	"<": "&lt;",
+	">": "&gt;",
+	"\"": "&quot;",
+	"'": "&#x27;"
+};
+
+// adapted from html-entities <https://github.com/mdevils/html-entities> (MIT LICENSE)
 export function htmlEncode(str, attribute) {
-	let res = str.replace(/&/g, "&amp;").
-		replace(/</g, "&lt;").
-		replace(/>/g, "&gt;");
-	if(attribute) {
-		res = res.replace(/"/g, "&quot;").
-			replace(/'/g, "&#x27;");
+	const rx = attribute ? /[&<>'"]/g : /[&<>]/g;
+	let res;
+	let match = rx.exec(str);
+	if(match) {
+		res = "";
```

Comment:
It used to be the case that string concatenation was expensive, so people collected strings in an array which was eventually `#join`d. We _might_ eke out a little more performance that way?

Reply:
I have seen so many unexpected results, that it would be always worth a try.
According to [The Performance of JavaScript String Concat](https://medium.com/@zhongdongy/the-performance-of-javascript-string-concat-e52466ca2b3a) form 2017, i expect that we are not gaining performance as long as we are not throwing a whole html document in one single `#encodeHtml` call.

But it would be explainable that `#join` will outperform at some input, when we make three assumptions about an engine:
* Array is fast.
* Strings are stored in a immutable piece of memory.
* Substrings are stored as a non owning reference to the original string.

Then just put the strings together will result in creating a new string, copy old content and deal with the old string which have to be deleted (or marked for it) on each concatination. Where on the `#join` strategy the result string will be created only once at the end, where the length of the underlying piece of memory is known.

Reply:
Hmm... when we take the medium post, then might the `#generateAttributes` will profit from using string concatination instead of `#join`, becuase it's a small amount of short strings. I make a try on that idea.

Reply:
> Hmm... when we take the medium post, then might the #generateAttributes will profit from using string concatination instead of #join, becuase it's a small amount of short strings. I make a try on that idea.

Bad idea from me. Raising the context will only cause more questions. Let's stick on `#encodeHtml`.

Reply:
It looks like string concatenation is indeed faster these days: https://jsben.ch/VBdEz - I'm not entirely sure how representative that sample really is, but it seems to be a negligible micro-optimization either way.